src/lightning/data/cache/reader.py

```diff
@@ -0,0 +1,186 @@
+# Copyright The Lightning AI team.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+from typing import Any, Dict, Optional, Tuple
+
+import numpy as np
+
+from lightning.data.cache.pytree import tree_unflatten, treespec_loads
+from lightning.data.cache.serializers import _SERIALIZERS, Serializer
+from lightning.data.datasets.env import _DistributedEnv
+
+
+class ChunksConfig:
+    def __init__(self, cache_dir: str, index_filenames: str):
+        self._cache_dir = cache_dir
+        self.index_filenames = sorted(index_filenames)
+        self._intervals = []
+        self._config = None
+        self._chunks = []
+
+        for filename in self.index_filenames:
+            with open(os.path.join(self._cache_dir, filename)) as f:
+                data = json.load(f)
+
+                if self._config is None:
+                    self._config = data["config"]
+
+                elif self._config != data["config"]:
+                    raise Exception("The config isn't consistent between chunks. This shouldn't have happened.")
+
+                self._chunks.extend(data["chunks"])
+
+        self._config["data_spec"] = treespec_loads(self._config["data_spec"])
+
+        for chunk in self._chunks:
+            start, end = chunk["interval"]
+            if (end - start) != chunk["samples"]:
+                raise Exception(
+                    "The config intervals doesn't match the number of samples. This shouldn't have happened."
+                )
+            self._intervals.append(chunk["interval"])
+
+        self._length = sum([chunk["samples"] for chunk in self._chunks])
+
+    @property
+    def intervals(self):
+        return self._intervals
+
+    @property
+    def data_format(self):
+        return self._config["data_format"]
+
+    @property
+    def config(self):
+        return self._config
+
+    def __getitem__(self, index: int) -> Tuple[str, int, int]:
+        """Find the associated chunk metadata."""
+        for interval_config, internal in enumerate(self._intervals):
```

Comment:
How bad is doing a linear scan each time we access an item?

Reply:
This would definitely have a cost. I am planing to write a binary search over the intervals or even pass the shard in from the Batch Sampler to avoid this altogether.

Reply:
Done !